File: LeetCode/1960. Maximum Product of the Length of Two Palindromic Substrings/solution.py
class Solution:
    def maxProduct(self, string: str) -> int:
        n = len(string)
        # print(n,s.count('a'))
        def manacher(s):
            sz = [1]*n 
            dp = [1]*n
            lo, hi = 0, 0
            for i in range(1,n):
                if i <= hi:
                    sz[i] = min(sz[lo+hi-i],hi-i+1)
                if i+sz[i]-1 >= hi:
                    while i+sz[i] < n and i-sz[i] >= 0 and s[i+sz[i]] == s[i-sz[i]]:
                        dp[i+sz[i]] = sz[i]*2+1
                        sz[i] += 1
                # print(i,lo,hi,sz[i])
                if i+sz[i]-1 > hi:
                    lo, hi = i-sz[i]+1, i+sz[i]-1
            return dp
        

        left = manacher(string)
        right = manacher(string[::-1])
        
        # print(right)
        for i in range(1,n):
            left[i] = max(left[i],left[i-1])
            right[i] = max(right[i],right[i-1])

        right = right[::-1]

        # print(left,right)
        ans = 0
        for i in range(n-1):
            ans = max(ans, left[i]*right[i+1])
        return ans





class Solution:
    def maxProduct(self, s: str) -> int:
        n = len(s)
        radius = [1]*n 
        left, right = 0, -1
        for i in range(n):
            if i <= right:
                radius[i] = min(radius[left+right-i], right-i+1)
            while i-radius[i] >= 0 and i+radius[i] < n and s[i-radius[i]] == s[i+radius[i]]:
                radius[i] += 1
            if i+radius[i]-1 > right:
                right = i+radius[i]-1
                left = i-radius[i]+1
        pre, suff = [1]*n, [1]*n 
        for i in range(n):
            j = i+radius[i]-1
            pre[j] = max(pre[j], radius[i]*2-1)
            j = i-radius[i]+1
            suff[j] = max(suff[j], radius[i]*2-1)
        for i in range(1,n):
            pre[i] = max(pre[i], pre[i-1])
        for i in range(n-1)[::-1]:
            pre[i] = max(pre[i], pre[i+1]-2)
        for i in range(n-1)[::-1]:
            suff[i] = max(suff[i], suff[i+1])
        for i in range(1,n):
            suff[i] = max(suff[i], suff[i-1]-2)
        return max(pre[i-1]*suff[i] for i in range(1,n))

File: LeetCode/1960. Maximum Product of the Length of Two Palindromic Substrings/solution.cpp
class Solution {
public:
    long long maxProduct(string s) {
        int n = s.size();
        vector<int> span(n);

        // manacher
        for (int i = 0, l = 0, r = -1; i < n; ++i) {
            span[i] = (i <= r ? min(span[l + r - i], r - i + 1) : 1);
            while (i - span[i] >= 0 && i + span[i] < n && s[i - span[i]] == s[i + span[i]]) {
                ++span[i];
            }
            if (i + span[i] - 1 > r) {
                l = i - span[i] + 1;
                r = i + span[i] - 1;
            }
        }

        vector<int> pre(n), suf(n);
        for (int i = 0; i < n; ++i) {
            pre[i + span[i] - 1] = max(pre[i + span[i] - 1], span[i] * 2 - 1);
            suf[i - span[i] + 1] = max(suf[i - span[i] + 1], span[i] * 2 - 1);
        }

        for (int i = n - 2; i >= 0; --i) {
            pre[i] = max(pre[i], pre[i + 1] - 2);
        }
        for (int i = 1; i < n; ++i) {
            pre[i] = max(pre[i], pre[i - 1]);
        }
        
        
        for (int i = 1; i < n; ++i) {
            suf[i] = max(suf[i], suf[i - 1] - 2);
        }
        for (int i = n - 2; i >= 0; --i) {
            suf[i] = max(suf[i], suf[i + 1]);
        }

        long long ans = 0;
        for (int i = 0; i < n - 1; ++i) {
            ans = max(ans, (long long)pre[i] * suf[i + 1]);
        }
        return ans;
    }
};


// 作者：zerotrac2
// 链接：https://leetcode.cn/problems/maximum-product-of-the-length-of-two-palindromic-substrings/solution/liang-ge-hui-wen-zi-zi-fu-chuan-chang-du-soyb/
// 来源：力扣（LeetCode）
// 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。


作者：zerotrac2
链接：https://leetcode-cn.com/problems/maximum-product-of-the-length-of-two-palindromic-substrings/solution/liang-ge-hui-wen-zi-zi-fu-chuan-chang-du-soyb/
来源：力扣（LeetCode）
著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。